File: library/xml-model/src/main/java/org/cru/godtools/xml/model/Tabs.java
package org.cru.godtools.xml.model;

import com.google.common.collect.ImmutableList;

import org.ccci.gto.android.common.util.XmlPullParserUtils;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;

import static org.cru.godtools.xml.Constants.XMLNS_CONTENT;
import static org.cru.godtools.xml.model.Tab.XML_TAB;

public final class Tabs extends Content {
    static final String XML_TABS = "tabs";

    @NonNull
    private List<Tab> mTabs = ImmutableList.of();

    private Tabs(@NonNull final Base parent) {
        super(parent);
    }

    @NonNull
    public List<Tab> getTabs() {
        return mTabs;
    }

    @NonNull
    static Tabs fromXml(@NonNull final Base parent, @NonNull final XmlPullParser parser)
            throws IOException, XmlPullParserException {
        return new Tabs(parent).parse(parser);
    }

    @NonNull
    private Tabs parse(@NonNull final XmlPullParser parser) throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, XMLNS_CONTENT, XML_TABS);
        parseAttrs(parser);

        // process any child elements
        final List<Tab> tabs = new ArrayList<>();
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }

            // process recognized nodes
            switch (parser.getNamespace()) {
                case XMLNS_CONTENT:
                    switch (parser.getName()) {
                        case XML_TAB:
                            tabs.add(Tab.fromXml(this, parser, tabs.size()));
                            continue;
                    }
                    break;
            }

            // skip unrecognized nodes
            XmlPullParserUtils.skipTag(parser);
        }
        mTabs = ImmutableList.copyOf(tabs);

        return this;
    }
}

File: library/xml-model/src/main/java/org/cru/godtools/xml/model/Tabs.kt
package org.cru.godtools.xml.model

import androidx.annotation.VisibleForTesting
import org.ccci.gto.android.common.util.xmlpull.skipTag
import org.cru.godtools.xml.XMLNS_CONTENT
import org.xmlpull.v1.XmlPullParser

class Tabs : Content {
    companion object {
        internal const val XML_TABS = "tabs"
    }

    val tabs: List<Tab>
    override val tips get() = tabs.flatMap { it.tips }

    @VisibleForTesting
    internal constructor(parent: Base, tabs: List<Tab>) : super(parent) {
        this.tabs = tabs
    }

    @OptIn(ExperimentalStdlibApi::class)
    internal constructor(parent: Base, parser: XmlPullParser) : super(parent, parser) {
        parser.require(XmlPullParser.START_TAG, XMLNS_CONTENT, XML_TABS)

        tabs = buildList {
            while (parser.next() != XmlPullParser.END_TAG) {
                if (parser.eventType != XmlPullParser.START_TAG) continue

                when (parser.namespace) {
                    XMLNS_CONTENT -> when (parser.name) {
                        Tab.XML_TAB -> add(Tab(this@Tabs, size, parser))
                        else -> parser.skipTag()
                    }
                    else -> parser.skipTag()
                }
            }
        }
    }
}
